enterprise/server/util/oci/oci.go

```diff
@@ -305,3 +320,171 @@ func (t *mirrorTransport) RoundTrip(in *http.Request) (out *http.Response, err e
+// and returns the blob's or manifest's content type and content length if found.
+// If you can fetch metadata for a manifest or blob, the actual contents should be in the CAS.
+//
+// The digest in this case contains a registry, a repository, and a hash.
+// For example, 'index.docker.io` as the registry, 'alpine` as the repository, and 'sha256:somesha256...' as the hash.
+//
+// Callers must ensure that the user fetching the OCI manifest or blob from the AC or CAS is authorized to access this OCI image.
+// For private OCI registries, this means making a HEAD request to the registry before fetching from the AC or CAS.
+func FetchBlobOrManifestMetadataFromCache(ctx context.Context, acc repb.ActionCacheClient, bsc bspb.ByteStreamClient, ocidigest ctrname.Digest, ociResourceType ocipb.OCIResourceType) (string, int64, error) {
+	arRN, err := ocidigestToACResourceName(ocidigest, ociResourceType)
+	if err != nil {
+		return "", 0, err
+	}
+	ar, err := cachetools.GetActionResult(ctx, acc, arRN)
+	if err != nil {
+		return "", 0, err
+	}
+
+	var blobMetadataCASDigest *repb.Digest
+	var blobCASDigest *repb.Digest
+	for _, outputFile := range ar.GetOutputFiles() {
+		switch outputFile.GetPath() {
+		case blobMetadataOutputFilePath:
+			blobMetadataCASDigest = outputFile.GetDigest()
+		case blobOutputFilePath:
+			blobCASDigest = outputFile.GetDigest()
+		default:
+			log.CtxErrorf(ctx, "unknown output file path '%s' in ActionResult for %s", outputFile.GetPath(), ocidigest)
+		}
+	}
+	if blobMetadataCASDigest == nil || blobCASDigest == nil {
+		return "", 0, fmt.Errorf("missing blob metadata digest or blob digest for %s", ocidigest)
+	}
+	blobMetadataRN := digest.NewCASResourceName(
+		blobMetadataCASDigest,
+		"",
+		repb.DigestFunction_SHA256,
+	)
+	blobMetadata := &ocipb.OCIBlobMetadata{}
+	err = cachetools.GetBlobAsProto(ctx, bsc, blobMetadataRN, blobMetadata)
+	if err != nil {
+		return "", 0, err
+	}
+	return blobMetadata.GetContentType(), blobMetadata.GetContentLength(), nil
+}
+
+func ocidigestToACResourceName(ocidigest ctrname.Digest, ociResourceType ocipb.OCIResourceType) (*digest.ACResourceName, error) {
+	hash, err := v1.NewHash(ocidigest.DigestStr())
+	if err != nil {
+		return nil, fmt.Errorf("could not parse hash in %s: %s", ocidigest.Context(), err)
+	}
+	arKey := &ocipb.OCIActionResultKey{
+		Registry:      ocidigest.Context().RegistryStr(),
+		Repository:    ocidigest.Context().RepositoryStr(),
+		ResourceType:  ociResourceType,
+		HashAlgorithm: hash.Algorithm,
+		HashHex:       hash.Hex,
+	}
+	arKeyBytes, err := proto.Marshal(arKey)
+	if err != nil {
+		return nil, err
+	}
+	arDigest, err := digest.Compute(bytes.NewReader(arKeyBytes), repb.DigestFunction_SHA256)
+	if err != nil {
+		return nil, err
+	}
+	arRN := digest.NewACResourceName(
+		arDigest,
+		arInstanceName(ocidigest.Context().RegistryStr(), ocidigest.Context().RepositoryStr()),
```

Comment:
Basing the instance name on the registry, repository, and a salt is the right thing to do once this code handles OCI images from private registries.
However, it does mean that the registry mirror will start over with a cold cache once this change is deployed. I'm not sure if that will lead to a burst of failed actions as the cache is populated post-deploy.